vowpalwabbit/core/include/vw/core/thread_pool.h

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+#pragma once
+
+#include "vw/core/memory.h"
+
+#include <atomic>
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+// thread pool based off of the simple pool implementation from ccia:
+// https://github.com/anthonywilliams/ccia_code_samples
+
+namespace VW
+{
+template <typename T>
+class thread_safe_queue

```

Comment:
This is really similar to `queue.h` can we reuse it?

Reply:
that one is a pointer queue though

Reply:
Pointers are movable so usage of that type can probably be replaced to one common implementation that satisfies both

Reply:
who owns the objects though?

Reply:
might need to change its name from ptr_queue

Reply:
I dont think pointers need special treatment. We should be able to reuse the same implementation